File: android/app/src/main/kotlin/com/gps/tracker/BatteryChannelHandler.kt
package com.gps.tracker

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

private const val CHANNEL = "battery_channel"

/**
 * Handles one-shot battery level requests from Flutter via [MethodChannel].
 *
 * Reads the current battery percentage using Android's [BatteryManager] —
 * no third-party battery packages are used.
 */
class BatteryChannelHandler(
    private val context: Context,
    private val messenger: BinaryMessenger
) : MethodChannel.MethodCallHandler {

    private val channel = MethodChannel(messenger, CHANNEL)

    /** Registers this handler on the method channel. */
    fun register() {
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getBatteryLevel" -> result.success(getBatteryLevel())
            else -> result.notImplemented()
        }
    }

    /**
     * Reads the battery level using [BatteryManager] on API 21+,
     * or falls back to the sticky broadcast on older devices.
     *
     * @return Battery percentage in range 0–100, or -1 on failure.
     */
    private fun getBatteryLevel(): Int {
        val batteryManager = context.getSystemService(Context.BATTERY_SERVICE) as? BatteryManager
        return batteryManager?.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY) ?: run {
            val intent = context.registerReceiver(
                null,
                IntentFilter(Intent.ACTION_BATTERY_CHANGED)
            )
            val level  = intent?.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) ?: -1
            val scale  = intent?.getIntExtra(BatteryManager.EXTRA_SCALE, -1) ?: -1
            if (level >= 0 && scale > 0) (level * 100 / scale) else -1
        }
    }
}

File: ios/Runner/BatteryChannelHandler.swift
import Flutter
import UIKit

private let kChannelName = "battery_channel"

/**
 * Handles one-shot battery level reads from Flutter via `MethodChannel`.
 *
 * Uses `UIDevice.current.batteryLevel` — no third-party battery packages are used.
 */
class BatteryChannelHandler: NSObject, FlutterStreamHandler {

    private let channel: FlutterMethodChannel

    init(messenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(name: kChannelName, binaryMessenger: messenger)
        super.init()
        channel.setMethodCallHandler(handle(_:result:))

        // Must enable battery monitoring before reading level.
        UIDevice.current.isBatteryMonitoringEnabled = true
    }

    // MARK: - MethodCallHandler

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getBatteryLevel":
            result(batteryLevel())
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Battery Reading

    /**
     * Returns the current battery level as an integer percentage (0–100).
     * Returns -1 if the level cannot be determined.
     */
    private func batteryLevel() -> Int {
        let level = UIDevice.current.batteryLevel
        guard level >= 0 else { return -1 }
        return Int(level * 100)
    }

    // MARK: - FlutterStreamHandler (unused here)

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? { nil }
    func onCancel(withArguments arguments: Any?) -> FlutterError? { nil }
}
